programs/mpl-core/src/processor/write_external_plugin_adapter_data.rs

```diff
@@ -25,12 +97,184 @@ pub(crate) struct WriteCollectionExternalPluginAdapterDataV1Args {
+        _ => return Err(MplCoreError::UnsupportedOperation.into()),
+    }
+
+    // AppData and LifecycleHook both write the data after the plugin.
+    // LinkedAppData writes the data to the asset directly.
+    match wrapped_plugin {
+        ExternalPluginAdapter::LifecycleHook(_) | ExternalPluginAdapter::AppData(_) => {
+            let header = header.ok_or(MplCoreError::PluginsNotInitialized)?;
+            let registry = registry.ok_or(MplCoreError::PluginsNotInitialized)?;
+            match (data, buffer) {
+                (Some(data), None) => update_external_plugin_adapter_data(
+                    record,
+                    Some(core),
+                    header,
+                    registry,
+                    account,
+                    payer,
+                    system_program,
+                    data,
+                ),
+                (None, Some(buffer)) => update_external_plugin_adapter_data(
+                    record,
+                    Some(core),
+                    header,
+                    registry,
+                    account,
+                    payer,
+                    system_program,
+                    &buffer.data.borrow(),
+                ),
+                (Some(_), Some(_)) => Err(MplCoreError::TwoDataSources.into()),
+                (None, None) => Err(MplCoreError::NoDataSources.into()),
+            }
+        }
+        ExternalPluginAdapter::LinkedAppData(app_data) => {
+            let (_, mut header, mut registry) =
+                create_meta_idempotent::<T>(account, payer, system_program)?;
+
+            match fetch_wrapped_external_plugin_adapter(
+                account,
+                Some(core),
+                &ExternalPluginAdapterKey::DataSection(LinkedDataKey::LinkedAppData(
+                    app_data.data_authority,
+                )),
+            ) {
+                Ok((section_record, _)) => match (data, buffer) {
+                    (Some(data), None) => update_external_plugin_adapter_data(
+                        &section_record,
+                        Some(core),
+                        &mut header,
+                        &mut registry,
+                        account,
+                        payer,
+                        system_program,
+                        data,
+                    ),
+                    (None, Some(buffer)) => update_external_plugin_adapter_data(
+                        &section_record,
+                        Some(core),
+                        &mut header,
+                        &mut registry,
+                        account,
+                        payer,
+                        system_program,
+                        &buffer.data.borrow(),
+                    ),
+                    (Some(_), Some(_)) => Err(MplCoreError::TwoDataSources.into()),
+                    (None, None) => Err(MplCoreError::NoDataSources.into()),
+                },
+                Err(_) => match (data, buffer) {
```

Comment:
Hmm wondering if just using `Err(_)` will lead to some strange issue if there are certain types of errors.  I think really what is expected here is `Err(MplCoreError::ExternalPluginAdapterNotFound)`, but what if it was some strange Borsh error or something?

Suggestions to consider: You could match on specific error, or make the error case you want into some kind of `Result<Option<>>`, i.e., the return `Ok(None)` shows it is not an error, but is an expected "not found" condition.

Reply:
Oh man this is such a good catch and would have been such a painful footgun. Fixed and thank you!